### ISSUE TITLE
Swift error checking for bad rings

### ISSUE THREAD

**Issue by timolow** [rcbops/rpc-openstack#800]
We need the ability to flag and remove hosts that have corrupt or faulty ring's.  We observed a proxy node exhibiting the following error after rebalance:

```
Jan 31 14:52:01 XXX-objectstorageinfra01_swift_proxy_container-XXXX proxy-server: ERROR Unhandled exception in request: #012
Traceback (most recent call last):#012
  File "/usr/local/lib/python2.7/dist-packages/swift/proxy/server.py", line 405, in handle_request#012
    return handler(req)#012
  File "/usr/local/lib/python2.7/dist-packages/swift/common/utils.py", line 2648, in wrapped#012
    return func(*a, **kw)#012  File "/usr/local/lib/python2.7/dist-packages/swift/proxy/controllers/base.py", line 263, in wrapped#012
    return func(*a, **kw)#012  File "/usr/local/lib/python2.7/dist-packages/swift/proxy/controllers/base.py", line 99, in wrapped#012
    return func(*a, **kw)#012  File "/usr/local/lib/python2.7/dist-packages/swift/proxy/controllers/obj.py", line 895, in PUT#012
    self.account_name, self.container_name, self.object_name)#012
  File "/usr/local/lib/python2.7/dist-packages/swift/common/ring/ring.py", line 308, in get_nodes#012
    part = self.get_part(account, container, obj)#012
  File "/usr/local/lib/python2.7/dist-packages/swift/common/ring/ring.py", line 259, in get_part#012
    self._reload()#012
  File "/usr/local/lib/python2.7/dist-packages/swift/common/ring/ring.py", line 157, in _reload#012
    ring_data = RingData.load(self.serialized_path)#012  File "/usr/local/lib/python2.7/dist-packages/swift/common/ring/ring.py", line 71, in load#012    magic = gz_file.read(4)#012  File "/usr/lib/python2.7/gzip.py", line 261, in read#012    self._read(readsize)#012  File "/usr/lib/python2.7/gzip.py", line 296, in _read#012    self._read_gzip_header()#012  File "/usr/lib/python2.7/gzip.py", line 190, in _read_gzip_header#012    raise IOError, 'Not a gzipped file'#012IOError: Not a gzipped file (txn: txXXXX) (client_ip: 10.X.X.X)
```

---
###### For each pull request, update the branch line to reference it.
- [x] master - https://github.com/rcbops/rpc-openstack/pull/1551
- [ ] mitaka-13.0
- [ ] liberty-12.2
- [ ] kilo
##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:
##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:
##### Release note
- [ ] This change includes a release note.
- [ ] This change does not require a release note.
##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues)
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc)
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by tomc603** [comment]
@timolow What would be an acceptable way to flag these hosts? Do you want removal to be automated, or do you want support to be notified of corruption so they can take corrective action manually?

**Comment by timolow** [comment]
The removal should be automated so we don't have fubar nodes in rotation, and notify us that a swift node is out of rotation because of corrupt rings.  If i recall the LB never marked the node down, be it a misconfiguration in our heathchecks or swift responds OK to /heathcheck with bad ring information.

**Comment by tomc603** [comment]
I've sent off an email to @timolow to ask some clarifying questions about this issue, its root cause, and if I could have a support ticket shared with me out of band so I can dig through a bit more.

Initially, I think this would have to be solved by watching log entries for specific kinds of exceptions, and reacting to them. As far as the load balancer issue goes, we can probably find a better check to perform, which would be able to catch issues like "HTTP is available, /healthcheck is fine, but reading data is impossible". I'm digging through that now, but I really need to better understand what led to the initial corruption so I can replicate it and investigate.

**Comment by tomc603** [comment]
Is there anybody else in @rcbops/support-feedback give input here? I'd be glad to forward the email I sent to @timolow if there is.

**Comment by TEKNISHEN** [comment]
Tim –

Can you provide the answers or additional detail for this bug submission?

Thanks,

SG

**Comment by tomc603** [comment]
@andymcc has suggested that we can create a swift static-web object, which is globally readable. Testing the access to this file will allow us to not only check that the proxy hosts are running, but also that they can access data in the ring. Performing this check from a load balancer (F5, haproxy, or any other) will allow us to remove failing systems from the load balancer automatically as well as triggering a MaaS alert.

Fully implementing this change will require that our load balancer checks be changed from TCP open port checks to full HTTP(s) checks. A status of "200 OK" should be indicative of a fully working swift stack returning data appropriately. However, load balancer checks that are not converted will continue to operate the same way they currently do, so changing these checks is only mandatory to gain the new functionality.

Additionally, by checking the entire access path for swift data, we should be able to create MaaS checks that are more correct. This should give Support a better insight into the customer experience, including the desired ability to verify customers can access data from their swift installations. A single check can replace all existing checks, because proxy status can be assumed when data is accessible, and proxy status can be inferred when specific errors (like a 500 response) are returned. 

Once this work is complete and functional, we should create documentation for customers utilizing their own load balancer(s) so they can create their own checks and configure any necessary monitoring outside of MaaS.

The task list should look like this:
* Create a new swift user for LB monitoring
* Add swift container and a small file to check
* Allow access to the test file
* Retrieve the AUTH_ URL for the test file
* Use the URL to configure monitoring
* Change LB monitors to use the URL in configured health checks

**Comment by tomc603** [comment]
Working on Ansible changes to create the needed Swift user and the check file. I've found so many ways in which Ansible _doesn't_ work...which is so nice.

**Comment by tomc603** [comment]
I'm refactoring some of this work, because I have realized installations not using MaaS may still want to perform this check. I'm also switching over to using the os_* modules from Ansible instead of using command-line commands and attempting to parse the response.

In my test environment, I was unable to fetch the index file in the static web without authentication, so that will require the loadbalancer check to authenticate while running the check.

**Comment by odyssey4me** [comment]
@tomc603 Be warned that the os_ modules from Ansible will require versions of python packages which are probably incompatible with what's deployed in RPC, and very probably a version of Ansible (2.x) that RPC can't use just yet. The older ones which support Ansible 1.x I hear are horrible - ymmv.

**Comment by tomc603** [comment]
@odyssey4me Thank you very much for that feedback! No doubt, you've saved me hours of tinkering. I just really hate the idea of parsing STDOUT in an ansible task, hoping that output doesn't change.

**Comment by b3rnard0** [comment]
@tomc603 Can you please provide an update for this card? Thanks.

**Comment by tomc603** [comment]
I have what appears to be a functional AIO environment where I'm squashing bugs on this work. I may have to rebuild it sooner rather than later due to some mistakes I made with Swift.

**Comment by tomc603** [comment]
More work has been done to clean up hard-coded items inside the task file. At this point, more testing is going to happen tomorrow (I blew away my test AIO and will create a new one), which should help debug anything that's broken, missing, or not working as expected.

Left to do is an F5 check for load balancing, and a MaaS http check for service status.

**Comment by tomc603** [comment]
AIO rebuild took roughly forever and failed a couple times. Should be making forward progress on this today.

**Comment by tomc603** [comment]
Manually fixed galera key for Mitaka so I could move ahead with testing.

**Comment by tomc603** [comment]
Continued testing and refinement of the swift file check deployment task. Even though I'm limiting my `when` clause to `groups['swift_proxy'][0]`, it is still selecting two hosts- The AIO host and a swift proxy container. Also, when a keystone function fails, subsequent tasks are skipped. So there's a bug in my logic at some point which will need to be teased out.

**Comment by tomc603** [comment]
The `swift_url_check` play is fully functional now. Remaining work is to use the registered URL fact to create a MaaS check, and an F5 check when appropriate.

**Comment by tomc603** [comment]
For some reason the PR isn't showing up here. It can be found https://github.com/rcbops/rpc-openstack/pull/1551

**Comment by tomc603** [comment]
This _may_ be blocked by https://github.com/rcbops/u-suk-dev/issues/475 due to the introduction of a new secret for a swift user in keystone for the purpose of creating a container and object in the container.

**Comment by tomc603** [comment]
Notes for those that may contribute in the future.

The approach to this issue has been to create a new swift container, and an object inside that container which can be checked by an F5 and/or MaaS checks. The thought goes a 200 means the test object can be accessed properly through proxies to the backing ring and back again. A 500 typically means the swift proxy is working properly, but the backing ring may not be. Obviously a 400 means the object is not present, which would probably indicate someone has deleted it incorrectly, or that it was not created and somehow the checks were created anyway.

After discussing this issue with Andy and Support, I think we've come to the agreement that Dev does not want to be in a position where we automatically remove swift rings, proxies, etc. from a topology. For this reason, we are not writing any tooling to do such a thing. If support feels it is okay to automate removing ring members or proxies, then they should be able to create a small wrapper to this work which meets their needs. But as the work stands now, a notification should be created for a non-200 return from swift, which support can handle as they see fit.

The object being used is a very simplistic HTML file, which is being uploaded into a new container with a configurable name. The swift user being created also has a configurable name and a new secret in the rpco secrets file. The swift user, container, and object are all created from the first swift node in the swift_proxies group, which sets a fact including the swift container's token. This fact is accessed by each member of swift_proxies when creating the MaaS remote HTTP check so we only have to search for that content once during setup. This seems like the simplest and cleanest approach to the problem rather than hitting swift with many requests unnecessarily during setup-maas.

Once this work is merged, we may be able to add an F5 check based on this work to determine if the entire swift topology is working properly, instead of making assumptions based on healthcheck status which may not be correct if ring data is corrupted. I would like to see this check running successfully for a while with few or no false alarms before attempting that work though. Creating customer outages due to inconsistent check results isn't something I'm looking to rush into.

**Comment by cloader89** [comment]
After looking into the issue, I need to fix https://github.com/rcbops/u-suk-dev/issues/475 first, as @git-harry suggested.

**Comment by alextricity25** [comment]
Yes, that issue needs to be fixed in order for the upgrade job to properly succeed on #1551. #1551  introduces a new secret variable, and right now we don't do any sort of migration of those new variables in ``test-upgrade.sh``, we just continue using the old ones. See https://github.com/rcbops/rpc-openstack/blob/master/scripts/test-upgrade.sh#L48-L59

**Comment by alextricity25** [comment]
Hey @tomc603. Thanks for making the fix! It looks like it just merged.
At the moment I don't think it would be best for use to backport this change over to the mitaka-13.1 and liberty-12.2 branches, and seeing that the other PR to migrate secrete variables was not backported, this won't work if it's backported. We are moving this over to done per our discussion during the standup today.
